lsm_tree.c

```diff
@@ -0,0 +1,333 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <stdbool.h>
+#include "lsm_tree.h"
+#include "memtable.h"
+#include "custom_io.h"
+
+// prototypes for static functions here
+static char** init_segments();
+static char* compact(char **segment_files);
+static char* merge_files(char *segment_file1, char *segment_file2);
+static void merge_lines(char *line_a, char *line_b, FILE *new_fp, bool *incr_a,
+bool *incr_b);
+static char* search_segments(char **segment_files, int full_segments, int key);
+static char* search_segment_file(FILE *segment_ptr, int key);
+static void free_segments(char **segments);
+
+/* Creates an LSM Tree for the program to use, initializing
+ * everything properly */
+LSM_Tree* init_lsm_tree() {
+	LSM_Tree *lsm_tree = (LSM_Tree*) malloc(sizeof(LSM_Tree));
+	if (lsm_tree == NULL) {
+		printf("Allocation of memory for LSM Tree failed.\n");
+		return NULL;
+	}
+
+	Memtable *memtable = init_memtable();
+	if (memtable == NULL) {
+		free(lsm_tree);
+		return NULL;
+	}
+
+	char **segments = init_segments();
+	if (segments == NULL) {
+		free(lsm_tree);
+		free(memtable);
+		return NULL;
+	}
+
+	lsm_tree->memtable = memtable;
+	lsm_tree->segments = segments;
+	lsm_tree->full_segments = 0;
+	return lsm_tree;
+}
+
+/* Handles the submission provided by user; returns 0 if all succeeds,
+ * otherwise returns -1.  */
+int handle_submission(LSM_Tree *lsm_tree, Submission *submission) {
+	if (submission->action == ADD) {
+		if (insert(lsm_tree->memtable, submission->key, submission->value)
+				== 0) {
+			lsm_tree->memtable->count_keys++;
+		} else {
+			printf("Insertion of new node failed.\n");
+			return -1;
+		}
+
+	} else if (submission->action == SEARCH) {
+		// search memtable first
+		MNode *node = search(lsm_tree->memtable, submission->key);
+		char *value;
+		if (node == NULL) {
+			// value wasn't in memtable, so look through the log files
+			value = search_segments(lsm_tree->segments, lsm_tree->full_segments,
+					submission->key);
+		} else {
+			value = node->data;
+		}
+
+		if (value != NULL) {
+			printf("The value for key %d is %s.\n", submission->key, value);
+		} else {
+			printf("Key not found in LSM Tree system.\n");
+		}
+
+	} else if (submission->action == DELETE) {
+		// always soft delete here; do not decrement keys in tree
+		if (delete(lsm_tree->memtable, submission->key, false) != 0) {
+			printf("Deletion of key, value pair with key %d failed.\n",
```

Comment:
np; for deletion the value should just be null so probably can just comment something along lines of ```Deletion of key failed```

Reply:
Changed wording to only mention key.